osu.Game.Rulesets.Taiko/Mods/TaikoModEnforceAlternate.cs

```diff
@@ -0,0 +1,162 @@
+        }
+
+        public void ApplyToHealthProcessor(HealthProcessor healthProcessor)
+        {
+            this.healthProcessor = healthProcessor;
+
+            if (CauseFail.Value)
+                healthProcessor.FailConditions += FailCondition;
+        }
+
+        protected bool FailCondition(HealthProcessor healthProcessor, JudgementResult result) => blockedKeystrokes > 0 && lastBlocked != null;
+
+        public void Update(Playfield playfield)
+        {
+            if (RepeatAfterDrumroll.Value)
+            {
+                if (currentRoll != null && currentRoll.AllJudged)
+                {
+                    if (currentRoll.HitObject.IsStrong)
+                    {
+                        blockedRim = null;
+                        blockedCentre = null;
+                    }
+                    else
+                        blockedCentre = null;
+
+                    currentRoll = null;
+                }
+
+                currentRoll = playfield.HitObjectContainer.Objects.ToList().Find(o => o is DrawableDrumRoll && !o.AllJudged) as DrawableDrumRoll;
+            }
+        }
+
+        private bool handleMultipleKeypresses(List<TaikoAction> rims, List<TaikoAction> centres)
+        {
+            bool bothLeft = (rims.Count == 1 && rims.First() == TaikoAction.LeftRim) && (centres.Count == 1 && centres.First() == TaikoAction.LeftCentre);
+            bool bothRight = (rims.Count == 1 && rims.First() == TaikoAction.RightRim) && (centres.Count == 1 && centres.First() == TaikoAction.RightCentre);
+
+            if (bothLeft || bothRight)
+            {
+                bool rimBlocked = rims.Any(a => a == blockedRim);
+                bool centreBlocked = centres.Any(a => a == blockedCentre);
+
+                if (rimBlocked && centreBlocked)
+                {
+                    blockedKeystrokes++;
+
+                    return true;
+                }
+                else
+                {
+                    blockedRim = rims.First();
+                    blockedCentre = centres.First();
+                }
+            }
+            else
+            {
+                if (rims.Count == 2)
+                    blockedRim = null;
+
+                if (centres.Count == 2)
+                    blockedCentre = null;
+            }
+
+            var blocked = rims.Concat(centres).ToList().FindAll(a => a == lastBlocked);
+
+            lastBlocked = null;
+
+            if (blocked.Any())
+                inputManager.RetractLastBlocked();
+
+            return false;
+        }
+
+        private bool handleBindings(TaikoAction? single, List<TaikoAction> pressedActions)
+        {
+            if (!healthProcessor.IsBreakTime.Value)
+            {
+                var rims = pressedActions.FindAll(a => a == TaikoAction.LeftRim || a == TaikoAction.RightRim);
+                var centres = pressedActions.FindAll(a => a == TaikoAction.LeftCentre || a == TaikoAction.RightCentre);
```

Comment:
Splitting this into lists feels awkward as you have at most 4 actions you care about at one given frame. Flag enums or a struct would be better and most likely faster. It would simplify a lot of the neighbouring code, too.